mixnet/mixnet.py

```diff
@@ -12,31 +16,105 @@
 class Mixnet:
     mix_nodes: List[MixNode]
 
-    # Build a new topology deterministically using an entropy.
-    # The entropy is expected to be injected from outside.
-    #
-    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
-    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
-    def build_topology(
+    def __init__(
         self,
-        entropy: bytes,
+        mix_nodes: List[MixNode],
         n_layers: int,
         n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+        initial_entropy: bytes,
+    ):
+        self.mix_nodes = mix_nodes
+
+        self.set_entropy(initial_entropy)
+
+        self.topology_updater = MixnetTopologyUpdater(
+            mix_nodes, n_layers, n_nodes_per_layer, mixnet_epoch_sec, proactive_sec
+        )
+        self.topology_updater.daemon = True
+        self.topology_updater.start()
+
+    def set_entropy(self, entropy: bytes) -> None:
+        """
+        Set entropy (seed) to the Fisher-Yates shuffler
+        that is going to be used for deterministic topology construction.
+
+        This method is expected to be called periodically by the user.
+        Until the next entropy is provides, Fisher-Yates shuffler continues to works
+        based on this entropy set once.
+        """
+        FisherYates.set_seed(entropy)
+
+    def choose_mixnode(self) -> MixNode:
+        return random.choice(self.mix_nodes)
+
+    def current_topology(self) -> MixnetTopology:
+        return self.topology_updater.current_topology()
+
+
+class MixnetTopologyUpdater(Thread):
+    def __init__(
+        self,
+        mix_nodes: List[MixNode],
+        n_layers: int,
+        n_nodes_per_layer: int,
+        mixnet_epoch_sec: int,
+        proactive_sec: int,
+    ):
+        super().__init__()
+        self.mix_nodes = mix_nodes
+        self.n_layers = n_layers
+        self.n_nodes_per_layer = n_nodes_per_layer
+        self.mixnet_epoch_sec = mixnet_epoch_sec
+        self.proactive_sec = proactive_sec
+
+        self.lock = threading.Lock()
+        self.topology = self.build_topology()
+        self.next_topology = None
+
+    def run(self) -> None:
+        next_epoch_ts = datetime.now() + timedelta(seconds=self.mixnet_epoch_sec)
+
+        while True:
+            time.sleep(1 / 1000)
+
+            now = datetime.now()
+            if now < next_epoch_ts - timedelta(seconds=self.proactive_sec):
```

Comment:
I hope that I understand this correctly, the intention here is to update the topology before the next epoch, so that when the new epoch starts it can already use the new topology?

If so, then it might be problematic due to the construction of the random beacon. That is, the beacon stability is more probable with time, that is, it is better to wait longer for the beacon. In other words, can we consider that the new topology must be usable with some delay, for example: the beacon is "stable" when it is in calculated in slot 1 of every epoch, then the topology is updated with some delay for example we assume that it is calculated in slot 2 and, usable for slot 3 and later of each epoch. Therefore, we are giving more time for network synchronization.

How does it sounds?

Also for me it is easier to reason when I we switch from thinking in seconds into thinking in slots and epochs. I see less synchronization problems ;)

Reply:
The strategy you suggested is reasonable if we design topology to be updated once per `consensus_epoch`. But, this PR assumed that the topology is updated once per `mixnet_epoch`, which is shorter than `consensus_epoch`, because I thought the `consensus_epoch` is going to be very long (e.g. 5days). I thought this is what we've discussed in Notion, but I probably didn't understand it correctly. Let me think which way between these two approach is more reasonable.

Reply:
Ah, even if we have `mixnet_epoch` shorter than `consensus_epoch`, we also need to use the newly injected entropy with some delays, as you explained.